arviz/data/io_pymc3.py

```diff
@@ -1,33 +1,80 @@
 """PyMC3-specific conversion code."""
 import logging
+from typing import Dict, List, Any, Optional, TYPE_CHECKING
+
 import numpy as np
 import xarray as xr
-
 from .. import utils
 from .inference_data import InferenceData
 from .base import requires, dict_to_dataset, generate_dims_coords, make_attrs
 
+if TYPE_CHECKING:
+    import pymc3 as pm
+
 _log = logging.getLogger(__name__)
 
+Coords = Dict[str, List[Any]]
+Dims = Dict[str, List[str]]
+
 
 class PyMC3Converter:
     """Encapsulate PyMC3 specific logic."""
 
+    model = None  # type: Optional[pm.Model]
+    nchains = None  # type: int
+    ndraws = None  # type: int
+    posterior_predictive = None  # Type: Optional[Dict[str, np.ndarray]]
+    prior = None  # Type: Optional[Dict[str, np.ndarray]]
+
     def __init__(
-        self, *, trace=None, prior=None, posterior_predictive=None, coords=None, dims=None
+        self,
+        *,
+        trace=None,
+        prior=None,
+        posterior_predictive=None,
+        coords: Optional[Coords] = None,
+        dims: Optional[Dims] = None,
+        model=None
     ):
+        import pymc3
+
+        self.pymc3 = pymc3
+
         self.trace = trace
+
         # This next line is brittle and may not work forever, but is a secret
         # way to access the model from the trace.
-        self.model = (
-            None
-            if trace is None
-            else self.trace._straces[0].model  # pylint: disable=protected-access
-        )
-        self.nchains = trace.nchains if hasattr(trace, "nchains") else 1
-        self.ndraws = len(trace)
+        if trace is not None:
+            self.model = self.trace._straces[0].model  # pylint: disable=protected-access
```

Comment:
should line 56 - 59 live up here, so that if `model` is provided, we do not use protected access?

Reply:
I wasn't sure, because I thought that might break previous code that was carelessly written inside a `with Model():` construct (which used to be useless, but also harmless).
I think it's the right thing, but didn't want to make that change without approval.
I'd encourage you to make that patch. If you prefer for me to do it, LMK.